File: kernel/src/video.rs
use spin::Mutex;

derive_kernel_enum! {
    pub enum LineType {
        HorizontalLine,
        VerticalLine,
    }
}

// From UEFI Specification
derive_kernel_enum! {
    pub enum EfiGraphicsPixelFormat {
        PixelRedGreenBlueReserved8BitPerColor,
        PixelBlueGreenRedReserved8BitPerColor,
        PixelBitMask,
    }
}

#[derive(Debug, Clone, Copy,)]
#[repr(C)]
pub struct BootVideoInfo {
  base_address: *mut u32,
  buffer_size: usize,
  width: u32,
  height: u32,
  pixels_per_scanline: u32,
  pixel_format: EfiGraphicsPixelFormat,
}

pub struct Color {
  pub r: u8,
  pub g: u8,
  pub b: u8,
}

unsafe impl Send for BootVideoInfo {}

impl BootVideoInfo {
  pub fn convert_pixel_format(&self, color: Color,) -> u32 {
    let format = self.pixel_format;
    match format
    {
      EfiGraphicsPixelFormat::PixelBlueGreenRedReserved8BitPerColor =>
      {
        ((color.b as u32) << 16) | ((color.g as u32) << 8) | (color.r as u32)
      }
      EfiGraphicsPixelFormat::PixelRedGreenBlueReserved8BitPerColor =>
      {
        ((color.r as u32) << 16) | ((color.g as u32) << 8) | (color.b as u32)
      }
      EfiGraphicsPixelFormat::PixelBitMask => 0,
    }
  }

  pub unsafe fn draw_pixel(&self, x: usize, y: usize, color: Color,) {
    let pixel = self.convert_pixel_format(color,);
    if x < self.width as usize && y < self.height as usize
    {
      let offset = y * (self.pixels_per_scanline as usize) + x;

      unsafe {
        self.base_address.add(offset,).write(pixel,);
      }
    }
  }

  pub unsafe fn fill_screen(&self, color: Color,) {
    let total_pixels = (self.pixels_per_scanline as usize) * (self.height as usize);
    let pixel = self.convert_pixel_format(color,);

    for i in 0..total_pixels
    {
      unsafe {
        self.base_address.add(i,).write(pixel,);
      }
    }
  }

  pub unsafe fn draw_line(
    &self,
    x: usize,
    y: usize,
    length: usize,
    color: Color,
    type_line: LineType,
  ) {
    let pixel = self.convert_pixel_format(color,);
    match type_line
    {
      LineType::HorizontalLine =>
      {
        let max_x = (x + length).min(self.width as usize,);
        for current_x in x..max_x
        {
          let offset = y * self.pixels_per_scanline as usize + current_x;
          unsafe {
            self.base_address.add(offset,).write(pixel,);
          }
        }
      }
      LineType::VerticalLine =>
      {
        let max_y = (y + length).min(self.height as usize,);
        for current_y in y..max_y
        {
          let offset = current_y * self.pixels_per_scanline as usize + x;
          unsafe {
            self.base_address.add(offset,).write(pixel,);
          }
        }
      }
    }
  }

  pub unsafe fn draw_rectangle(
    &self,
    x: usize,
    y: usize,
    width: usize,
    height: usize,
    color: Color,
  ) {
    let max_x = (x + width).min(self.width as usize,);
    let max_y = (y + height).min(self.height as usize,);
    let pixel = self.convert_pixel_format(color,);

    for current_y in y..max_y
    {
      for current_x in x..max_x
      {
        unsafe {
          let offset = current_y * (self.pixels_per_scanline as usize) + current_x;
          self.base_address.add(offset,).write(pixel,);
        }
      }
    }
  }
}

pub static FRAMEBUFFER: Mutex<Option<BootVideoInfo,>,> = Mutex::new(None,);

pub fn draw_pixel_global(x: usize, y: usize, color: Color,) {
  if let Some(fb,) = FRAMEBUFFER.lock().as_ref()
  {
    unsafe {
      fb.draw_pixel(x, y, color,);
    }
  }
}

pub fn fill_screen_global(color: Color,) {
  if let Some(fb,) = FRAMEBUFFER.lock().as_ref()
  {
    unsafe {
      fb.fill_screen(color,);
    }
  }
}

pub fn draw_line_global(x: usize, y: usize, length: usize, color: Color, type_line: LineType,) {
  if let Some(fb,) = FRAMEBUFFER.lock().as_ref()
  {
    unsafe {
      fb.draw_line(x, y, length, color, type_line,);
    }
  }
}

pub fn draw_rectangle_global(x: usize, y: usize, width: usize, height: usize, color: Color,) {
  if let Some(fb,) = FRAMEBUFFER.lock().as_ref()
  {
    unsafe {
      fb.draw_rectangle(x, y, width, height, color,);
    }
  }
}

File: kernel/video.c
#include "microkernel.h"

#include <video.h>

static const boot_video_info_t *g_video = NULL;

typedef enum {
  PIXEL_RED_GREEN_BLUE_RESERVED_8BIT_PER_COLOR,
  PIXEL_BLUE_GREEN_RED_RESERVED_8BIT_PER_COLOR,
  PIXEL_BIT_MASK,
  PIXEL_BLT_ONLY,
  PIXEL_FORMAT_MAX
} efi_graphics_pixel_format_t;

// --- Functions ---
// It is forward declaration used for init only in main function
void init_video(const boot_video_info_t *video) {
  g_video = video;
}

void draw_pixel(size_t x, size_t y, uint32_t color) {
  if (!g_video)
    return;

  if (x < g_video->width && y < g_video->height) {
    g_video->base_address[y * g_video->pixels_per_scan_line + x] = color;
  }
}

void fill_screen(uint32_t color) {
  if (!g_video)
    return;

  size_t total_pixels = (size_t)g_video->pixels_per_scan_line * g_video->height;
  for (size_t i = 0; i < total_pixels; i++) {
    g_video->base_address[i] = color;
  }
}

void draw_line(size_t x, size_t y, size_t length, uint32_t color, line_type_t type_line) {
  if (!g_video)
    return;

  if (type_line == LINE_HORIZONTAL) {
    size_t max_x = (x + length < g_video->width) ? x + length : g_video->width;
    for (size_t cx = x; cx < max_x; cx++) {
      draw_pixel(cx, y, color);
    }
  } else if (type_line == LINE_VERTICAL) {
    size_t max_y = (y + length < g_video->height) ? y + length : g_video->height;
    for (size_t cy = y; cy < max_y; cy++) {
      draw_pixel(x, cy, color);
    }
  }
}

void draw_rectangle(size_t x, size_t y, size_t width, size_t height, uint32_t color) {
  if (!g_video)
    return;

  size_t max_x = (x + width < g_video->width) ? x + width : g_video->width;
  size_t max_y = (y + height < g_video->height) ? y + height : g_video->height;

  for (size_t cy = y; cy < max_y; cy++) {
    for (size_t cx = x; cx < max_x; cx++) {
      draw_pixel(cx, cy, color);
    }
  }
}
